### PR TITLE
Slip10redo

### PR DESCRIPTION
**Pull request by wjuan-mob**
Rebase of [chris's slip10 PR ](https://github.com/mobilecoinfoundation/mobilecoin/pull/1183/#issuecomment-1027109039) onto master combined with [James's Mnmeonic pr] ( https://github.com/mobilecoinfoundation/mobilecoin/pull/780)

### CODE REVIEW

File: util/keyfile/src/config.rs [hunk 1]
```diff
@@ -1,48 +1,69 @@
...
+type VecBytes = Vec<u8>;
 /// Configuration for generating key files for a new user identity
 #[derive(Debug, StructOpt)]
 pub struct Config {
-    /// Optional FogURL for the accounts
-    #[structopt(short, long)]
-    pub acct: Option<String>,
+    /// Fog Report URL
+    #[structopt(long)]
+    pub fog_report_url: Option<String>,
```
Comment (remoun) [inline util/keyfile/src/config.rs:14]: Optional: Use `mc_util_uri::FogUri` to validate the URL format

File: util/keyfile/src/config.rs [hunk 2]
```diff
@@ -1,48 +1,69 @@
 // Copyright (c) 2018-2021 The MobileCoin Foundation
 
-use rand::{rngs::StdRng, SeedableRng};
-use std::path::PathBuf;
+use std::{cmp, fs, path::PathBuf};
+
 use structopt::StructOpt;
 
+// Hack to work around Vec special handling in structopt
+type VecBytes = Vec<u8>;
```
Comment (remoun) [inline util/keyfile/src/config.rs:8]: Is this still necessary when you specify `parse()` in the structop/clap args?

**Review by jcape — CHANGES_REQUESTED** [review]
Still some remaining instances of slip10 that should get removed---we should also treat the mnemonic string as a string, rather than doing weird things with bytes :-)

File: util/from-random/src/lib.rs [hunk 3]
```diff
@@ -13,3 +13,11 @@ pub trait FromRandom: Sized {
     /// Using a mutable RNG, take it's output to securely initialize the object
     fn from_random<R: CryptoRng + RngCore>(csprng: &mut R) -> Self;
 }
+
+impl FromRandom for [u8; 32] {
+    fn from_random<R: CryptoRng + RngCore>(csprng: &mut R) -> [u8; 32] {
+        let mut result = [0u8; 32];
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply util/from-random/src/lib.rs]: Good idea.

File: util/keyfile/src/bin/main.rs [hunk 4]
```diff
@@ -4,18 +4,37 @@
...
     };
-    let acct_key = AccountKey::from(&root_id);
-    println!("{:?}\n{:?}", root_id, acct_key,);
+
+    println!("{:?}", acct_key);
+}
+fn main() {
+    let mut n_files = 0usize;
+    for file_str in env::args() {
+        print_keyfile_bytes(fs::read(file_str).expect("Could not read file"));
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply util/keyfile/src/bin/main.rs]: Done.

File: util/keyfile/src/bin/keygen_main.rs [hunk 5]
```diff
@@ -2,29 +2,59 @@
...
+use rand::{RngCore, SeedableRng};
+use rand_hc::Hc128Rng;
 use structopt::StructOpt;
+#[derive(Debug, StructOpt)]
+struct Config {
+    #[structopt(flatten)]
+    pub general: GeneralConfig,
 
+    pub name: String,
+}
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply util/keyfile/src/bin/keygen_main.rs]: Done.

File: util/keyfile/src/error.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+impl From<AccountKeyError> for Error {
+    fn from(src: AccountKeyError) -> Error {
+        Error::AccountKey(src)
+    }
+}
+
+impl From<DecodeError> for Error {
+    fn from(src: DecodeError) -> Error {
+        Error::Decode(format!("{}", src))
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply util/keyfile/src/error.rs:46]: I think the format approach is fine. So I'll leave it as is. I agree that it's equivalent.

File: util/keyfile/src/config.rs (hunk 2 above)
Comment (wjuan-mob) [reply util/keyfile/src/config.rs]: I think so? Maybe I was doing it wrong, but it was complaining when I was using Vec<u8> directly.

File: util/keyfile/src/config.rs (hunk 1 above)
Comment (wjuan-mob) [reply util/keyfile/src/config.rs:14]: I'll do this separately if it's still desired.

File: fog/sample-paykit/proto/remote_wallet.proto [hunk 7]
```diff
@@ -32,8 +32,8 @@ service RemoteWalletApi {
 }
 
 message FreshBalanceCheckRequest {
-    /// Root entropy for the account the new client will be using.
-    bytes root_entropy = 1;
+    /// Mnemonic key for the account the new client will be using.
+    bytes mnemonic = 1;
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply fog/sample-paykit/proto/remote_wallet.proto]: Done.

File: mobilecoind/strategies/accounts.py [hunk 8]
```diff
@@ -37,8 +37,8 @@ def connect(host, port):
 def register_account(key_data, stub) -> AccountData:
     # Generate an account key from this root entropy
     resp = stub.GetAccountKeyFromRootEntropy(
-        mobilecoind_api_pb2.GetAccountKeyFromRootEntropyRequest(
-            root_entropy=bytes(key_data['root_entropy'])))
+        mobilecoind_api_pb2.GetAccountKeyFromMnemonic(
+            mnemonic=bytes.fromhex(key_data['mnemonic'])))
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply mobilecoind/strategies/accounts.py]: Done.

File: tools/fog-local-network/fog_conformance_tests.py [hunk 9]
```diff
@@ -154,7 +154,7 @@ def start(self):
         with grpc.insecure_channel(self.remote_wallet_host_port) as channel:
             stub = remote_wallet_pb2_grpc.RemoteWalletApiStub(channel)
             response = self._retrying_grpc_request(stub.FreshBalanceCheck, remote_wallet_pb2.FreshBalanceCheckRequest(
-                root_entropy=bytes(key['root_entropy']),
+                mnemonic=bytes.fromhex(key['mnemonic']),
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply tools/fog-local-network/fog_conformance_tests.py]: Done.

File: util/keyfile/src/json_format.rs [hunk 10]
```diff
@@ -1,22 +1,95 @@
...
+use mc_crypto_rand::{CryptoRng, RngCore};
+use mc_util_from_random::FromRandom;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// JSON schema for a slip10 identity
+#[derive(Clone, PartialEq, Eq, Hash, Default, Debug, Serialize, Deserialize)]
+pub struct Slip10IdentityJson {
+    /// Slip10 key
+    #[serde(serialize_with = "as_hex", deserialize_with = "from_hex")]
+    pub slip10_key: [u8; 32],
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply util/keyfile/src/json_format.rs]: Done.

File: util/keyfile/src/backwards_compatibility.json [hunk 11]
```diff
@@ -0,0 +1 @@
... jelly column manage grit panda quiz assume modify error faith never broccoli","account_index":0,"fog_report_url":null,"fog_report_id":null,"fog_authority_spki":null},{"mnemonic":"music display finger thrive aspect genius obtain reason jar chimney someone brisk resemble pulp coast spawn border much focus autumn logic pave win liberty","account_index":0,"fog_report_url":null,"fog_report_id":null,"fog_authority_spki":null},{"mnemonic":"broccoli venture whale lizard antique title leader sick hedgehog skirt until happy this furnace admit siege man lawn toe enforce elder negative suggest treat","account_index":0,"fog_report_url":null,"fog_report_id":null,"fog_authority_spki":null},{"mnemonic":"buffalo strike film school celery embody cinnamon spot night uncover monitor useless uncle eager purpose brother essence lion vital doll cattle transfer dynamic turtle","account_index":0,"fog_report_url":null,"fog_report_id":null,"fog_authority_spki":null},{"mnemonic":"ethics wash coil hello never harsh one brand side chaos cause blood course orange write beach seat fat park enforce feed reward airport cement","account_index":0,"fog_report_url":null,"fog_report_id":null,"fog_authority_spki":null}]
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply util/keyfile/src/backwards_compatibility.json:1]: I just copied this from your PR. I hadn't actually bothered to check these. I assumed that they were standard.

File: util/keyfile/src/lib.rs [hunk 12]
```diff
@@ -107,90 +122,102 @@ fn to_io_error<E: 'static + std::error::Error + Send + Sync>(err: E) -> std::io:
...
+    use mc_account_keys_slip10::{Slip10Key, Slip10KeyGenerator};
 
-impl std::error::Error for ProstError {
-    fn description(&self) -> &str {
-        "prost deserialization failed"
+    /// Test that round-tripping through a keyfile without fog gets the same
+    /// result as creating the key directly.
+    #[test]
+    fn keyfile_roundtrip_no_fog() {
+        let dir = tempfile::tempdir().expect("Could not create temp dir");
```
Comment (jcape) [inline util/keyfile/src/lib.rs:137]: This isn't for this PR, but we should make a test utils crate that uses `TempDir::new_in()` to put all these into the `OUT_DIR`, and replace all our localized `tempfile` users.

Comment (wjuan-mob) [reply util/keyfile/src/lib.rs:137]: Created a ticket for this.

File: util/keyfile/src/bin/main.rs [hunk 13]
```diff
@@ -5,18 +5,38 @@
...
-    let acct_key = AccountKey::from(&root_id);
-    println!("{:?}\n{:?}", root_id, acct_key,);
+
+    println!("{:?}", acct_key);
+}
+
+fn main() {
+    let mut n_files = 0usize;
+    for path in env::args().skip(1) {
+        print_keyfile_bytes(fs::read(path).expect("Could not read file"));
```
Comment (remoun) [inline util/keyfile/src/bin/main.rs:31]: Let's include the path
```suggestion
        print_keyfile_bytes(fs::read(path).expect(&format!("Could not read file '{}'", path)));
```

File: util/keyfile/src/bin/main.rs [hunk 14]
```diff
@@ -5,18 +5,38 @@
...
-            None => mc_util_keyfile::read_keyfile_data(&mut std::io::stdin())
-                .unwrap_or_else(|_| panic!("Failed when reading from stdin")),
-            Some(arg) => mc_util_keyfile::read_keyfile(arg)
-                .unwrap_or_else(|_| panic!("Failed when reading from {}", arg)),
-        }
+fn print_keyfile_bytes(bytes: Vec<u8>) {
+    let acct_key = if let Ok(identity) =
+        mc_util_keyfile::read_root_entropy_keyfile_data(Cursor::new(bytes.as_slice()))
+    {
+        println!("{:?}", identity);
```
Comment (remoun) [inline util/keyfile/src/bin/main.rs:18]:
```suggestion
        println!("Identity: {:?}", identity);
```

**Review by remoun — APPROVED** [review]
Rust & proto code LGTM. I'm not familiar with the Python code, but the changes look reasonable.

File: Cargo.lock [hunk 15]
```diff
@@ -6303,14 +6309,37 @@ dependencies = [
...
 
+[[package]]
+name = "prost"
+version = "0.9.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "444879275cb4fd84958b1a1d5420d15e6fcf7c235fe47f053c9c2a80aceb6001"
+dependencies = [
+ "bytes 1.1.0",
+ "prost-derive 0.9.0",
+]
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply Cargo.lock]: Done.

File: util/keyfile/src/bin/keygen_main.rs [hunk 16]
```diff
@@ -3,29 +3,60 @@
...
 //! A CLI tool for generating individual MobileCoin identities
 
+use bip39::{Language, Mnemonic};
 use clap::Parser;
-use mc_account_keys::{RootEntropy, RootIdentity};
-use mc_util_keyfile::config::Config;
+use mc_util_keyfile::{config::Config as GeneralConfig, keygen};
+use rand::{RngCore, SeedableRng};
+use rand_hc::Hc128Rng;
+#[derive(Debug, Parser)]
```
Comment (remoun) [inline util/keyfile/src/bin/keygen_main.rs:11]: Nit: Add empty line and rustdoc.

```suggestion

/// Keygen config.
#[derive(Debug, Parser)]
```

File: util/keyfile/src/bin/keygen_main.rs [hunk 17]
```diff
@@ -3,29 +3,60 @@
...
+struct Config {
+    #[clap(flatten)]
+    pub general: GeneralConfig,
 
+    /// The key name.
+    pub name: String,
+}
```
Comment (remoun) [inline util/keyfile/src/bin/keygen_main.rs:18]: Nit: Separate with newline
```suggestion
}

```

File: util/keyfile/src/error.rs [hunk 18]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! This module contains the error type which methods of the `keyfile` crate can
+//! generate.
+
+use crate::mnemonic_acct::Error as MnemonicAccountError;
+use displaydoc::Display;
+use mc_account_keys::Error as AccountKeyError;
+use mc_account_keys_slip10::Error as Slip10Error;
+use mc_util_serial::DecodeError as MCUtilSerialDecodeError;
```
Comment (remoun) [inline util/keyfile/src/error.rs:10]: This is just a re-export of `prost::DecodeError`, so let's remove it

**Review by remoun — APPROVED** [review]

File: util/keyfile/src/bin/main.rs [hunk 19]
```diff
@@ -5,18 +5,38 @@
...
-                .unwrap_or_else(|_| panic!("Failed when reading from {}", arg)),
-        }
+fn print_keyfile_bytes(bytes: Vec<u8>) {
+    let acct_key = if let Ok(identity) =
+        mc_util_keyfile::read_root_entropy_keyfile_data(Cursor::new(bytes.as_slice()))
+    {
+        println!("Identity: {:?}", identity);
+        AccountKey::from(&identity)
+    } else {
+        mc_util_keyfile::read_keyfile_data(Cursor::new(bytes.as_slice()))
```
Comment (jcape) [inline util/keyfile/src/bin/main.rs:21]:
```suggestion
fn print_keyfile_bytes(bytes: &[u8]) {
    let acct_key = if let Ok(identity) =
        mc_util_keyfile::read_root_entropy_keyfile_data(Cursor::new(bytes))
    {
        println!("Identity: {:?}", identity);
        AccountKey::from(&identity)
    } else {
        mc_util_keyfile::read_keyfile_data(Cursor::new(bytes))
```

File: util/keyfile/src/error.rs (hunk 18 above)
Comment (jcape) [reply util/keyfile/src/error.rs:10]: Micro-nit: rust tends to use strict CamelCase names for types (e.g. `JsonError` and `IoError` below).

File: util/keyfile/src/lib.rs [hunk 20]
```diff
@@ -3,63 +3,79 @@
...
 pub mod keygen;
 
-use json_format::RootIdentityJson;
-use mc_account_keys::{PublicAddress, RootIdentity};
+use crate::error::Error;
+use bip39::Mnemonic;
+use mc_account_keys::{AccountKey, PublicAddress, RootIdentity};
 use mc_api::printable::PrintableWrapper;
-use std::{convert::TryInto, fs::File, io::prelude::*, path::Path};
+use std::{convert::TryInto, fs::File, io, io::prelude::*, path::Path};
```
Comment (jcape) [inline util/keyfile/src/lib.rs:18]:
```suggestion
use std::{convert::TryInto, fs::File, io::Read, path::Path};
```

**Review by jcape — COMMENTED** [review]
I think we can skip the io::prelude, no? Like, only `std::io::Read` is actually used?

File: util/keyfile/src/lib.rs (hunk 20 above)
Comment (wjuan-mob) [reply util/keyfile/src/lib.rs]: Also needed io:write since we implement the trait. Made this change.

File: util/keyfile/src/bin/main.rs (hunk 19 above)
Comment (wjuan-mob) [reply util/keyfile/src/bin/main.rs]: Done.

**Review by jcape — APPROVED** [review]

**Pull request merged**